The output of the tests so far:

`claytonh@claytonh-ThinkPad-X240:~/pycoal/tests$ nosetests -v mineral_test.py
mineral_test.test_normalize ... ok
mineral_test.test_indexOfGreaterThan ... FAIL
mineral_test.test_trainClassifier ... ok
mineral_test.test_saveClassifier ... ok
mineral_test.test_readClassifier ... ok
mineral_test.test_classifyPixel ... ok
mineral_test.test_classifyImage ... ok
mineral_test.test_classifyImages ... ok

======================================================================
FAIL: mineral_test.test_indexOfGreaterThan
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/claytonh/pycoal/tests/mineral_test.py", line 19, in test_indexOfGreaterThan
    assert test_classifier._MineralClassification__indexOfGreaterThan(test_list, 11) == 9
AssertionError

----------------------------------------------------------------------
Ran 8 tests in 0.002s

FAILED (failures=1)
`

--- a/pycoal/mineral.py
+++ b/pycoal/mineral.py
@@ -53,7 +53,8 @@ class MineralClassification:
         # return normalized image
         return normalizedImage
 
-    def __indexOfGreaterThan(self, elements, value):
+    @staticmethod
+    def __indexOfGreaterThan(elements, value):
         """
         Return the index of the first element greater than the value.
         Args:

--- a/pycoal/mineral.py
+++ b/pycoal/mineral.py
@@ -1,4 +1,14 @@
-#!/usr/bin/env python
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
 
 import spectral
 import pickle
@@ -25,7 +35,7 @@ class MineralClassification:
 
         # constrain wavelength range
         if wavelengthList is None:
-            wavelengthList = image.metadata.get('wavelength')
+            wavelengthList = map(float, image.metadata.get('wavelength'))
 
         # get the indices of the minimum and maximum element
         i = self.__indexOfGreaterThan(wavelengthList, self.wavelength_min)
@@ -55,7 +65,8 @@ class MineralClassification:
 
         for index,element in enumerate(elements):
             if element > value:
-                return index
+                break
+        return index
 
     def trainClassifier(self, libraryFileName, classifierType='gaussian'):
         """
@@ -68,7 +79,8 @@ class MineralClassification:
                                                       perceptron, gaussian, or
                                                       mahalanobisdistance.
         Returns:
-           classifier (PerceptronClassifier): trained classifier
+           classifier (PerceptronClassifier, GaussianClassifier,
+                       or MahalanobisDistanceClassifier): trained classifier
         """
 
         # open the digital spectral library
@@ -165,13 +177,13 @@ class MineralClassification:
         # open the image
         image = spectral.io.aviris.open(imageFilename)
 
+        # get wavelength list
+        wavelengthList = map(float, image.metadata.get('wavelength'))
+
         # change units from nanometers to micrometers if applicable
         if image.metadata.get('wavelength units') == u'Nanometers':
             image.metadata['wavelength'] = [x / 1000 for x in image.metadata.get('wavelength')]
             image.metadata['wavelength units'] = u'Micrometers'
-
-        # get wavelength list
-        wavelengthList = image.metadata.get('wavelength')
 
         # allocate an MxN array for the classified image
         classifiedImage = numpy.ndarray((image.shape[0] * image.shape[1],))

--- a/pycoal/mineral.py
+++ b/pycoal/mineral.py
@@ -94,10 +94,7 @@ class MineralClassification:
         if classifierType == 'perceptron':
             # try out using the neural net structure used in the following
             # paper: http://www.aaai.org/Papers/FLAIRS/1999/FLAIRS99-057.pdf
-            nn_structure = numpy.ndarray((3,))
-            nn_structure[0] = 9
-            nn_structure[1] = 35
-            nn_structure[2] = 60
+            nn_structure = [9, 35, 60]
             classifier = spectral.classifiers.PerceptronClassifier(nn_structure)
         elif classifierType == 'gaussian':
             classifier = spectral.classifiers.GaussianClassifier()

--- a/tests/mineral_test.py
+++ b/tests/mineral_test.py
@@ -1,4 +1,14 @@
-#!/usr/bin/env python
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
 
 from pycoal import mineral
 
@@ -6,17 +16,31 @@ def test_normalize():
     pass
 
 def test_indexOfGreaterThan():
+    """
+    This tests the private method __indexOfGreaterThan in the class MineralClassification
+    Args:
+        None
+    Returns:
+        None
+    """
     # test object
     test_classifier = mineral.MineralClassification()
 
+    test_list = [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
+
     # [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
-    test_list = [x * 1.0 for x in range(10)]
+    #                         ---  ---
+    # 6.0 is the first value greater than 5.0, so the return should be its index (6)
+    assert test_classifier._MineralClassification__indexOfGreaterThan(test_list, 5.0) == 6
 
-    assert test_classifier._MineralClassification__indexOfGreaterThan(test_list, 5) == 6
+    # [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
+    #     ---  ---
+    # 2.0 is the first value greater than 1.0, so the return should be its index (2)
+    assert test_classifier._MineralClassification__indexOfGreaterThan(test_list, 1.0) == 2
 
-    assert test_classifier._MineralClassification__indexOfGreaterThan(test_list, 1) == 2
-
-    assert test_classifier._MineralClassification__indexOfGreaterThan(test_list, 11) == 9
+    # [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
+    # 11.0 is not in the range of this list, so the return will just be the last index
+    assert test_classifier._MineralClassification__indexOfGreaterThan(test_list, 11.0) == 9
 
 def test_trainClassifier():
     pass

--- a/tests/mineral_test.py
+++ b/tests/mineral_test.py
@@ -25,24 +25,21 @@ def test_indexOfGreaterThan():
     Returns:
         None
     """
-    # test object
-    test_obj = mineral.MineralClassification()
-
     test_list = [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
 
     # [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
     #                         ---  ---
     # 6.0 is the first value greater than 5.0, so the return should be its index (6)
-    assert test_obj._MineralClassification__indexOfGreaterThan(test_list, 5.0) == 6
+    assert mineral.MineralClassification._MineralClassification__indexOfGreaterThan(test_list, 5.0) == 6
 
     # [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
     #     ---  ---
     # 2.0 is the first value greater than 1.0, so the return should be its index (2)
-    assert test_obj._MineralClassification__indexOfGreaterThan(test_list, 1.0) == 2
+    assert mineral.MineralClassification._MineralClassification__indexOfGreaterThan(test_list, 1.0) == 2
 
     # [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
     # 11.0 is not in the range of this list, so the return will just be the last index
-    assert test_obj._MineralClassification__indexOfGreaterThan(test_list, 11.0) == 9
+    assert mineral.MineralClassification._MineralClassification__indexOfGreaterThan(test_list, 11.0) == 9
 
 def test_trainClassifier():
     pass

--- a/tests/mineral_test.py
+++ b/tests/mineral_test.py
@@ -11,45 +11,125 @@
 # limitations under the License.
 
 from pycoal import mineral
+import spectral
+import os
 
 def test_normalize():
     pass
 
 def test_indexOfGreaterThan():
     """
-    This tests the private method __indexOfGreaterThan in the class MineralClassification
+    This tests the private method __indexOfGreaterThan in the class MineralClassification.
     Args:
         None
     Returns:
         None
     """
     # test object
-    test_classifier = mineral.MineralClassification()
+    test_obj = mineral.MineralClassification()
 
     test_list = [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
 
     # [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
     #                         ---  ---
     # 6.0 is the first value greater than 5.0, so the return should be its index (6)
-    assert test_classifier._MineralClassification__indexOfGreaterThan(test_list, 5.0) == 6
+    assert test_obj._MineralClassification__indexOfGreaterThan(test_list, 5.0) == 6
 
     # [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
     #     ---  ---
     # 2.0 is the first value greater than 1.0, so the return should be its index (2)
-    assert test_classifier._MineralClassification__indexOfGreaterThan(test_list, 1.0) == 2
+    assert test_obj._MineralClassification__indexOfGreaterThan(test_list, 1.0) == 2
 
     # [0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
     # 11.0 is not in the range of this list, so the return will just be the last index
-    assert test_classifier._MineralClassification__indexOfGreaterThan(test_list, 11.0) == 9
+    assert test_obj._MineralClassification__indexOfGreaterThan(test_list, 11.0) == 9
 
 def test_trainClassifier():
     pass
 
 def test_saveClassifier():
-    pass
+    """
+    This tests the public method saveClassifier in the class MineralClassification.
+    Args:
+        None
+    Returns:
+        None
+    """
+    # test object
+    test_obj = mineral.MineralClassification()
+
+    # the files the test will write to
+    test_dump_files = ["perceptron.p", "gaussian.p", "mahalanobi.p"]
+
+    # create arbitrary structure for perceptron classifier
+    test_struct = [9, 35, 60]
+
+    # perceptron classifier test object
+    test_perceptron = spectral.classifiers.PerceptronClassifier(test_struct)
+
+    # dump classifier to file
+    test_obj.saveClassifier(test_perceptron, test_dump_files[0])
+
+    # assert classifier was dumped to specified file in test_dump_files
+    assert os.path.isfile(test_dump_files[0])
+
+    # gaussian classifier test object
+    test_gaussian = spectral.classifiers.GaussianClassifier()
+
+    # dump classifier to file
+    test_obj.saveClassifier(test_gaussian, test_dump_files[1])
+
+    # assert classifier was dumped to specified file in test_dump_files
+    assert os.path.isfile(test_dump_files[1])
+
+    # mahalanobi's distance classifier test object
+    test_mahalanobi = spectral.classifiers.MahalanobisDistanceClassifier()
+
+    # dump classifier to file
+    test_obj.saveClassifier(test_mahalanobi, test_dump_files[2])
+
+    # assert classifier was dumped to specified file in test_dump_files
+    assert os.path.isfile(test_dump_files[2])
+
 
 def test_readClassifier():
-    pass
+    """
+    This tests the public method readClassifier in the class MineralClassification.
+    Args:
+        None
+    Returns:
+        None
+    """
+    # test object
+    test_obj = mineral.MineralClassification()
+
+    # the files the test will read from
+    test_load_files = ["perceptron.p", "gaussian.p", "mahalanobi.p"]
+
+    # load perceptron classifier
+    test_perceptron = test_obj.readClassifier(test_load_files[0])
+
+    # assert the object is loaded as the PerceptronClassifier type
+    assert type(test_perceptron) == spectral.classifiers.PerceptronClassifier
+
+    # assert the shape is equivalent to the shape specified in test_saveClassifier
+    assert test_perceptron.shape == [9, 35, 60]
+
+    # load gaussian classifier
+    test_gaussian = test_obj.readClassifier(test_load_files[1])
+
+    # assert the object is loaded as the GaussianClassifier type
+    assert type(test_gaussian) == spectral.classifiers.GaussianClassifier
+
+    # load mahalanobi's distance classifier
+    test_mahalanobi = test_obj.readClassifier(test_load_files[2])
+
+    # assert the object is loaded as the GaussianClassifier type
+    assert type(test_mahalanobi) == spectral.classifiers.MahalanobisDistanceClassifier
+
+    # clean up after save and read tests
+    for f in test_load_files:
+        os.remove(f)
 
 def test_classifyPixel():
     pass

